src/combinatorics/additive/behrend.lean

```diff
@@ -0,0 +1,198 @@
+integer points on that sphere and map them onto `ℕ` in a way that preserves arithmetic progressions
+(`behrend.map`).
+
+## Main declarations
+
+* `behrend.sphere`: The intersection of the Euclidean sphere with the positive integer quadrant.
+  This is the set that we will map on `ℕ`.
+* `behrend.map`: The map from the Euclidean space to `n`. It reads off the coordinates as digits in
+  base `d`.
+* `behrend.card_sphere_le_roth_number_nat`: Implicit lower bound on Roth numbers in terms of
+  `behrend.sphere`.
+
+## References
+
+* [Bryan Gillespie, *Behrend’s Construction*]
+  (http://www.epsilonsmall.com/resources/behrends-construction/behrend.pdf)
+* Behrend, F. A., "On sets of integers which contain no three terms in arithmetical progression"
+* [Wikipedia, *Salem-Spencer set*](https://en.wikipedia.org/wiki/Salem–Spencer_set)
+
+## Tags
+
+Salem-Spencer, Behrend construction, arithmetic progression, sphere, strictly convex
+-/
+
+open finset nat real
+open_locale big_operators pointwise
+
+namespace behrend
+variables {α β : Type*} {n d k : ℕ} {x : fin n → ℕ}
+
+/-!
+### Turning the sphere into a Salem-Spencer set
+
+We define `behrend.sphere`, the intersection of the $$L^2$$ sphere with the positive quadrant of
+integer points. Because the $$L^2$$ closed ball is strictly convex, the $$L^2$$ sphere and
+`behrend.sphere` are Salem-Spencer (`add_salem_spencer_sphere`). Then we can turn this set in
+`fin n → ℕ` into a set in `ℕ` using `behrend.map`, which preserves `add_salem_spencer` because it is
+an additive monoid homomorphism.
+-/
+
+/-- The box `{0, ..., d - 1}^n` as a finset. -/
+def box (n d : ℕ) : finset (fin n → ℕ) := fintype.pi_finset $ λ _, range d
+
+lemma mem_box : x ∈ box n d ↔ ∀ i, x i < d := by simp only [box, fintype.mem_pi_finset, mem_range]
+
+@[simp] lemma card_box : (box n d).card = d ^ n := by simp [box]
+@[simp] lemma box_zero : box (n + 1) 0 = ∅ := by simp [box]
+
+/-- The intersection of the sphere of radius `sqrt k` with the integer points in the positive
+quadrant. -/
+def sphere (n d k : ℕ) : finset (fin n → ℕ) := (box n d).filter $ λ x, ∑ i, x i^2 = k
+
+lemma sphere_zero_subset : sphere n d 0 ⊆ 0 :=
+λ x, by simp [sphere, function.funext_iff] {contextual := tt}
+
+@[simp] lemma sphere_zero_right (n k : ℕ) : sphere (n + 1) 0 k = ∅ := by simp [sphere]
+
+lemma sphere_subset_box : sphere n d k ⊆ box n d := filter_subset _ _
+
+lemma norm_of_mem_sphere {n d k : ℕ} {x : fin n → ℕ} (hx : x ∈ sphere n d k) :
+  ∥(pi_Lp.equiv 2 _).symm (coe ∘ x : fin n → ℝ)∥ = sqrt k :=
+begin
+  rw euclidean_space.norm_eq,
+  dsimp,
+  simp_rw [norm_coe_nat, ←cast_pow, ←cast_sum, (mem_filter.1 hx).2],
+end
+
+/-- The map that appears in Behrend's bound on Roth numbers. -/
+@[simps] def map (d : ℕ) : (fin n → ℕ) →+ ℕ :=
+{ to_fun := λ a, ∑ i, a i * d ^ (i : ℕ),
+  map_zero' := by simp_rw [pi.zero_apply, zero_mul, sum_const_zero],
+  map_add' := λ a b, by simp_rw [pi.add_apply, add_mul, sum_add_distrib] }
```

Comment:
If you wanted, could golf to 
```suggestion
def map (d : ℕ) : (fin n → ℕ) →+ ℕ :=
∑ i : fin n, (add_monoid_hom.mul_right (d ^ (i : ℕ))).comp (pi.eval_add_monoid_hom _ i)

@[simp] lemma map_apply (d : ℕ) (a : fin n → ℕ) : map d a = ∑ i, a i * d ^ (i : ℕ) :=
add_monoid_hom.finset_sum_apply _ _ _
```

Reply:
Yeah, it breaks some stuff. Given that this construction is local to the proof, I won't spend effort into fixing it.